### ISSUE TITLE
Module initialization order

### ISSUE THREAD

**Issue by ric2016** [fisharebest/webtrees#3073]
In some of my custom modules, I have to replace certain webtrees classes, such as GedcomRecord. This isn't nice but generally works by modifying the class loader from the custom module.

However, classes that are already in use cannot be replaced (e.g. a recent modification of SiteMapModule.php causes GedcomRecord to be initialized there, apparently).

Can we load custom modules prior to core modules, in order to give them a chance to carry out such replacements? This would make webtrees more customizable.

This would be a simple change in `ModuleService.php` lines 540-541 (function 'all'), swapping the two function calls:

`return $this->customModules()
                ->merge($this->coreModules())`

**Comment by fisharebest** [comment]
> isn't nice but generally works by modifying the class loader

:-)

I think you would have the same problem if another custom module used (and hence loaded) that class before you...

> swapping the two function calls:

That order was intentional.  Firstly, I thought that custom modules should be able to rely on the core modules being available/initialised.  Secondly, I thought it would make it harder for custom modules to break the application.

I will admit that I didn't think in too much detail about this, and all the consequences.
But at the moment, I would prefer not to change it.

I *think* the correct solution is to have a `GedcomRecordFactory` object, which is responsible for creating `GedcomRecord` objects.

e.g. replace

```
$indivdiual = Individual::getInstance(...);
```

with

```
$factory = app(GedcomRecordFactory::class);
$individual = $factory->getIndividualInstance(...);
```

We would also need to declare `IndividualInterface`, etc.

Then, you could register your own factory, which would create your own objects (which implement the Interface).

I'm not sure how easy it would be to convert all the code to this structure.

**Comment by fisharebest** [comment]
> I'm not sure how easy it would be to convert all the code to this structure.

Perhaps not *too* difficult.

The existing `getInstance()` functions would simply need to be replaced with calls to the factory object.

**Comment by fisharebest** [comment]
This is actually a change that I've been thinking about for a while, as it would make testing easier.

**Comment by ric2016** [comment]
GedcomRecord was just one example. But I generally agree that it would be preferable to modify the webtrees core code, so that custom modules won't need these replacements in the first place.

Regarding the actual GedcomRecord extension, I use an approch where factories for specific types may be registered (so I guess it is overall a GedcomRecordFactoryFactory), like this:

`GedcomRecordExt::addFactory('_LOC', new SharedPlaceFactory(...));`

(where SharedPlaceFactory is a GedcomRecordFactory. See [here](https://github.com/vesta-webtrees-2-custom-modules/vesta_shared_places/tree/master/patchedWebtrees) for the code)

This is probably similar to what you have in mind. If webtrees would allow something like this, it would be very useful for me indeed!

**Comment by fisharebest** [comment]
> This is probably similar to what you have in mind.

Yes - that sort of thing.

**Comment by ric2016** [comment]
See pull request for a minimal change. This would help for my use case - Not sure how useful it would be in general for testing, i.e. the use case you mentioned.

**Comment by ric2016** [comment]
I'll create another PR for this based on the suggestions in #3085. I'd really like this to make it into the next release - otherwise some of my custom modules will break.

**Comment by fisharebest** [comment]
Can you give me a link to your code that creates modified gedcom records.

I am sure you already shared this - but I cannot find it.

**Comment by ric2016** [comment]
The code from the custom module is [here](https://github.com/vesta-webtrees-2-custom-modules/vesta_shared_places/tree/master/patchedWebtrees). (The pull request contains a similar implementation. It will have to be modified based on [these suggestions](https://github.com/fisharebest/webtrees/issues/3085#issuecomment-587137916). We should probably also use `app('cache.array')` to cache the gedcom records?)

**Comment by fisharebest** [comment]
FYI, the core code now includes a submitter type.  You should add it.

https://github.com/vesta-webtrees-2-custom-modules/vesta_shared_places/blob/master/patchedWebtrees/GedcomRecordExt.php#L140

**Comment by ric2016** [comment]
Thanks for the note - I have also adjusted GedcomRecordFactories in the updated [PR](https://github.com/fisharebest/webtrees/pull/3091).

**Comment by fisharebest** [comment]
Can you take a look at the develop branch?

It has an initial version of GedcomRecord factories.
Note: I just finished coding this.
It isn't tested or reviewed.
It is late and I will continue tomorrow...

My thoughts are:

You can extend `GedcomRecordFactory` to create new record types like this.

```
class CustomGedcomRecordFactory extends GedcomRecordFactory {
    protected function newGedcomRecord(string $type, string $xref, string $gedcom, ?string $pending, Tree $tree): ?GedcomRecord
    {
        if ($type === '_LOC') {
            return new SharedLocation($xref, $gedcom, $pending, $tree);
        }

        return parent::newGedcomRecord($type, $xref, $gedcom, $pending, $tree);
    }
}
```

You can register your factory in your module's `boot()` method.

```
function boot()
{
    $cache = app('cache.array');

    app()->instance(GedcomRecordFactoryInterface::class, new CustomGedcomRecordFactory($cache));
}
```

I think this approach will give you the features you need.  Let me know.

**Comment by ric2016** [comment]
This works for me. I can add factories for new types, and it's also possible to replace specific factories directly, which is nice.

I'm still not sure about the caching: It may be more flexible this way (each specific factory decides how to cache for itself), but the caching isn't optional functionally, at least for INDI and FAM (disabling the cache in the respecive factories leads to "Nesting level too deep - recursive dependency?" errors). Should this be addressed differently?

We should also ensure that instances are always created though the factories, this isn't the case yet.
(e.g. in IndividualPage.php Line 248 - that's just one occurence which I noted when testing a custom IndividualFactory which changes the name display).

I will continue to test this.

**Comment by fisharebest** [comment]
The "nesting level too deep" error was occuring in a call to `array_unique()`.

Individual had links to families, and the families had links to the individual.

When these aren't the same objects, PHP was stuck in a loop.

I have fixed this.  It now works without a cache - but is obviously much slower.

**Comment by ric2016** [comment]
Resolved by changes in webtrees 2.0.4.
